### PR TITLE
Adds supermatter attackby_hit typecache to deal with similarly-handled objects

### PR DESCRIPTION
**Pull request by admeeer**
## About The Pull Request

Adds a typecache for objects that are similarly handled in the attackby_hit proc

I decided to omit cigarette from the list since it has so much custom code. I could just check to see if the item is a cigarette, but I don't think that is very readable.

## Why It's Good For The Game

Makes interaction more efficient, also because a maintainer said so

## Changelog

:cl:
code: made an eensy teensie weensie change to some supermatter boilerplate
/:cl:

### CODE REVIEW

File: code/datums/components/supermatter_crystal.dm [hunk 1]
```diff
@@ -4,6 +4,8 @@
 	var/datum/callback/tool_act_callback
 	///Callback used by the SM to get the damage and matter power increase/decrease
 	var/datum/callback/consume_callback
+	// Typecache for similarly handled objects in attackby_proc
+	var/list/attackby_hit_typecache
```
Comment (timothymtorres) [inline code/datums/components/supermatter_crystal.dm:8]:
```suggestion
	// A whitelist of items that can interact with the SM without dusting the user
	var/static/list/sm_item_whitelist = typecacheof(list(
		/obj/item/melee/roastingstick,
		/obj/item/toy/crayon/spraycan,
	))
```

Try this instead.  You want your list to be properly named and for it to be static. (meaning it won't ever change)

File: code/datums/components/supermatter_crystal.dm [hunk 2]
```diff
@@ -22,6 +24,11 @@
 	src.tool_act_callback = tool_act_callback
 	src.consume_callback = consume_callback
 
+	attackby_hit_typecache = typecacheof(list(
+		/obj/item/melee/roastingstick,
+		/obj/item/toy/crayon/spraycan
+	))
```
Comment (timothymtorres) [inline code/datums/components/supermatter_crystal.dm:30]:
```suggestion

```

Delete this and move this to the `/datum/component/supermatter_crystal` location.  We shouldn't be creating this list when a SM spawns. (it should be pre-existing)

File: code/datums/components/supermatter_crystal.dm [hunk 3]
```diff
@@ -152,9 +159,7 @@
 	var/atom/atom_source = source
 	if(!istype(item) || (item.item_flags & ABSTRACT) || !istype(user))
 		return
-	if(istype(item, /obj/item/melee/roastingstick))
-		return FALSE
-	if(istype(item, /obj/item/toy/crayon/spraycan))
+	if(is_type_in_typecache(item, attackby_hit_typecache))
```
Comment (timothymtorres) [inline code/datums/components/supermatter_crystal.dm:162]:
```suggestion
	if(is_type_in_typecache(item, sm_item_whitelist))
```

File: code/datums/components/supermatter_crystal.dm (hunk 1 above)
Comment (admeeer) [reply code/datums/components/supermatter_crystal.dm]: yeah that makes sense, I should of thought of that

File: code/datums/components/supermatter_crystal.dm (hunk 2 above)
Comment (admeeer) [reply code/datums/components/supermatter_crystal.dm]: also makes sense

File: code/datums/components/supermatter_crystal.dm (hunk 3 above)
Comment (admeeer) [reply code/datums/components/supermatter_crystal.dm]: done

**Comment by timothymtorres** [comment]
Also this would be considered code improvement and not a refactor.  Refactors are usually when you dissect the code and rebuild it to be better.  It usually takes a lot longer to do, like hours instead of minutes.

There is no hard criteria when it comes to differentiating the two, but a few lines of code like this PR is a small amount of effort.

**Comment by jlsnow301** [comment]
what is going on with your branch

**Pull request merged**